lib/common/rules.c

```diff
@@ -483,43 +483,78 @@ evaluate_lt(const xmlNode *date_expression, const char *id,
 
 /*!
  * \internal
- * \brief Evaluate a date_expression
+ * \brief Evaluate a rule's date expression for a given date/time
  *
- * \param[in]  expr         XML of rule expression
- * \param[in]  now          Time to use for evaluation
- * \param[out] next_change  If not NULL, set to when evaluation will change
+ * \param[in]     date_expression  XML of a PCMK_XE_DATE_EXPRESSION element
+ * \param[in]     now              Time to use for evaluation
+ * \param[in,out] next_change      If not NULL, set to when evaluation will
+ *                                 change
  *
- * \return Standard Pacemaker return code
+ * \return Standard Pacemaker return code (unlike most other evaluation
+ *         functions, this can return either pcmk_rc_ok or pcmk_rc_within_range
+ *         on success)
```

Comment:
Correct on all counts, especially the last one ;)

The `pcmk__check_rules()` -> crm_rule -> pcs chain distinguishes OK from within range, though I don't remember the specifics anymore

Reply:
> The `pcmk__check_rules()` -> crm_rule -> pcs chain distinguishes OK from within range, though I don't remember the specifics anymore

via `pcmk_rc2exitc()`, which maps `pcmk_rc_within_range` to `CRM_EX_OK

Reply:
The update to address review doesn't map `pcmk_rc_within_range` to `pcmk_rc_ok` before returning from `pcmk__evaluate_date_expression()`. Not sure if you wanted to do that now, later, or never -- but sounds like not "never."

Reply:
Thinking further, I believe we could deprecate `pcmk_rc_within_range` as redundant and avoid the whole little headache. As of this PR, it looks like the only place where we treat it differently from `pcmk_rc_ok` is in selecting a text output message in `rule_check_default()`:
```
    switch (result) {
        case pcmk_rc_within_range:
            return out->info(out, "Rule %s is still in effect", rule_id);
        case pcmk_rc_ok:
            return out->info(out, "Rule %s satisfies conditions", rule_id);
```

It's not unreasonable to have only `pcmk_rc_{before,after}_range` and `pcmk_rc_ok`.

Reply:
Hmm, I thought there was more to it. I'll leave it just in case we're overlooking something, and for backward compatibility, but it does look like it could change.

Reply:
@clumens , do you remember why we added pcmk_rc_within_range instead of just using pcmk_rc_ok? That was a long time ago now ...

Reply:
I have a vague memory of wanting to be able to differentiate the idea of "this code exited successfully" from a value being in the expected range.  Or perhaps we just liked the symmetry of before/in/after.  Maybe the code that cared is no longer with us - like, perhaps the function could have returned pcmk_rc_ok for some other reason and it no longer does.

Reply:
> differentiate the idea of "this code exited successfully" from a value being in the expected range

I could see that. The annoying thing in practice is that this only applies to one type of rule expression, and we have to merge it with the results of evaluating the other expressions into the RC for the whole rule. To me, it's awkward that the rule's "OK" code is `pcmk_rc_within_range` if it contains a date expression and `pcmk_rc_ok` otherwise.

> perhaps the function could have returned pcmk_rc_ok for some other reason and it no longer does.

Very plausible